apparmor/README.md

```diff
@@ -1,103 +1,487 @@
-  to regain access to it with an `owner` statement in the same profile
+In this lab you will learn the basics of AppArmor and how to use it with Docker for improved security.
 
-  - For file operations, `r` corresponds to read, `w` to write, `k` to lock, `l` to link, and `x` to execute
+You will complete the following steps as part of this lab.
 
- This should get you started fairly well with AppArmor, but for more information you can consult the official [AppArmor documentation wiki](http://wiki.apparmor.net/index.php/Documentation) (under active development at this time).
+- [Step 1 - AppArmor primer](#primer)
+- [Step 2 - The `default-docker` AppArmor profile](#default-docker)
+- [Step 3 - Running a Container without an AppArmor profile](#no-profile)
+- [Step 4 - AppArmor and defense in depth](#depth)
+- [Step 5 - Custom AppArmor profile](#custom)
+- [Step 6 - Extra for experts](#extra)
 
-## Default AppArmor in Docker
+# Prerequisites
 
-1.  We can view the status of AppArmor on our Docker host by running `apparmor_status` (may require admin credentials) -- you should notice the `docker-default` profile is in enforce mode.  As described above, this is Docker's default AppArmor profile that is applied to containers on `docker run`.
+You will need all of the following to complete this lab:
 
-2.  To prove that this profile is applied by default, run an alpine container in another terminal `docker run --rm -it alpine sh` and then check `apparmor_status` again.  You should now see that a process (our alpine container!) has a profile defined and is also in enforce mode:
-    ```
-    1 processes have profiles defined.
-    1 processes are in enforce mode.
-       docker-default (28462)
-    ```
+- A Linux-based Docker Host with AppArmor enabled in the kernel
+- Docker 1.12 or higher
+
+The following command shows you how to check if AppArmor is enabled in your system's kernel:
+
+   Check from Docker 1.12 or higher
+   ```
+   $ docker version | grep apparmor
+   Security Options: apparmor seccomp   
+   ```
+   If the above output does not return a line with `apparmor` then your system does not have AppArmor enabled in its kernel.
+
+
+# <a name="primer"></a>Step 1: AppArmor primer
+
+By default, Docker applies the `docker-default` AppArmor profile to new containers. This profile is located in `/etc/apparmor.d/docker/` and you can find more information about it in the [documentation](https://docs.docker.com/engine/security/apparmor/#understand-the-policies).  
+
+Here are some quick pointers for how to understand AppArmor profiles:
+
+  - `Include` statements, such as `#include <abstractions/base>`, behave just like their `C` counterparts by expanding to additional AppArmor profile contents.
+
+  - AppArmor `deny` rules are evaluated first and have precedence over `allow` and `owner` rules. This means that deny rules cannot be overridden by subsequent allow or owner rules for the same resource.
```

Comment:
For those familiar with `allow, deny` type configurations, we should note that an allow will be overridden by a subsequent deny on the same resource. That confused me for a while when I first used AppArmor. I tried to do an `allow owner` followed by `deny everything`.